### PR TITLE
Submission embargo fix by removing group select

### PR DESCRIPTION
**Pull request by benbosman**
## References
* REST part of https://github.com/DSpace/dspace-angular/issues/869 (Does not fully fix issue though)
* Related to [REST Contract](https://github.com/DSpace/Rest7Contract/pull/152)

## Description
This is the REST part to fix the embargo bug https://github.com/DSpace/dspace-angular/issues/869
It removes the group select as discussed in the issue and REST contract

## Instructions for Reviewers
Keep in mind for this change, the Angular part is also required. The REST contract details how the access conditions can be defined

List of changes in this PR:
* The group selection has been removed from the spring config, java classes, REST, …
* When setting an access condition, the parameters are now also validated first (are the dates correct, …)
* The solution to create the resource policies based on the access condition has been centralized to the AccessConditionOption class
* This also removes the patchUploadAddAccessConditionTest IT because that test was specific to the group selection

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/WorkspaceItemRestRepositoryIT.java [hunk 1]
```diff
@@ -4666,4 +4610,380 @@ public void findOneFullProjectionTest() throws Exception {
...
+            .andExpect(jsonPath("$.sections.upload.files[0].accessConditions[0].name", is("openaccess")));
+
+        // verify that access conditions have been persisted
+        getClient(authToken)
+            .perform(get("/api/submission/workspaceitems/" + wItem.getID()))
+            .andExpect(status().isOk())
+            .andExpect(jsonPath("$.sections.upload.files[0].accessConditions[0].name", is("openaccess")));
+    }
+
+    // TODO: all cases in 74236 that start with "upload a new bitstream"
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/WorkspaceItemRestRepositoryIT.java:4677]: I'm not sure what this TODO is referring to? Maybe this was an internal comment/reminder?

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/WorkspaceItemRestRepositoryIT.java [hunk 2]
```diff
@@ -4666,4 +4610,380 @@ public void findOneFullProjectionTest() throws Exception {
...
+        getClient(authToken)
+            .perform(get("/api/submission/workspaceitems/" + wItem.getID()))
+            .andExpect(status().isOk())
+            .andExpect(jsonPath("$.sections.upload.files[0].accessConditions", empty()));
+    }
+
+    // The REST endpoint doesn't seem to support uploading a bitstream with the access conditions
+    @Test
+    @Ignore
+    public void uploadBitstreamWithAccessConditionOpenAccess() throws Exception {
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/WorkspaceItemRestRepositoryIT.java:4646]: I wonder if we should just remove this ignored test.  If I recall correctly, we had planned for this to be a two-step process. You upload the Bitstream, then restrict it. So, I think this scenario is already covered by the next method `patchBitstreamWithAccessConditionOpenAccess()`, which does just that...uploads the file, then restricts it.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/WorkspaceItemRestRepositoryIT.java [hunk 3]
```diff
@@ -4666,4 +4610,380 @@ public void findOneFullProjectionTest() throws Exception {
...
+
+        // verify that access conditions array is still empty
+        getClient(authToken)
+            .perform(get("/api/submission/workspaceitems/" + wItem.getID()))
+            .andExpect(status().isOk())
+            .andExpect(jsonPath("$.sections.upload.files[0].accessConditions", empty()));
+    }
+
+    @Test
+    public void patchBitstreamWithAccessConditionLease() throws Exception {
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/WorkspaceItemRestRepositoryIT.java:4953]: Could we rename this or describe this test better?  It wasn't immediately obvious to me why setting a Lease should throw an error, but I expect that's cause there's no date set?  If so, this should probably be renamed `patchBitstreamWithAccessConditionLeaseMissingDate` or similar to make it clear why it should fail.

Either that or add comments inline on why the error is expected.

**Review by tdonohue — COMMENTED** [review]
@benbosman : Did a quick code review today, and this looks good & is in line with what I expected.  I have a few minor things to cleanup in the ITs (nothing big at all though).  I haven't had a chance to test this yet, but will try and find time tomorrow.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@benbosman : I've tested this today & it works as described in the corresponding REST Contract PR https://github.com/DSpace/RestContract/pull/152.  However, my code review feedback above needs to be resolved, so I'm submitting this as "Request changes"

I'll also note that this PR _alone_ does NOT seem to solve the Angular issue https://github.com/DSpace/dspace-angular/issues/869 .  It appears a corresponding Angular PR will be needed, as the empty Group selectbox still appears.  So, I've updated the description of this PR to remove the `Fixes` keyword...as that keyword would cause the Angular issue ticket to be closed when this PR is merged.

**Comment by abollini** [comment]
hi @tdonohue thanks to ping me on this issue, I admit to haven't noted the issue nor the corresponding REST PR.
As you noted in the issue this feature was currently available in previous versions so it is a pity that we are now removing it instead than fix... anyway if we are going to do that please be sure to create a task to reconsider this decision.
Indeed, the real fix would be not complicated, instead than expose the uuid of the parent group having the (angular) client to discover the sub-groups we would have modified the access condition configuration exposed via REST to directly list the group and the uuid. The patch operation could have been extended to check against illegal attempt to select unexpected group. I will take a look to the code in the next days as I understood that other edge condition and fixes are provided included so this could be a first PR regardless to a reconsideration of the decision to drop the feature

**Comment by tdonohue** [comment]
@abollini : I've responded to you over in your more detailed comment at https://github.com/DSpace/dspace-angular/issues/869#issuecomment-773478173  Simply put, I don't think we have time to rethink this now, but I've created a follow-up ticket at https://github.com/DSpace/dspace-angular/issues/1011 where we can discuss how best to restore this feature.

So, for the purposes of this PR, I'd recommend we stay with the current decision & remove the feature for 7.0.  We can then discuss how to properly restore it (and ensure we can [resolve the issues noted in the original analysis](https://github.com/DSpace/dspace-angular/issues/869#issuecomment-713594267)).  Thanks!

**Review by abollini — CHANGES_REQUESTED** [review]
thanks for this PR @benbosman As you have invested time in adding additional test for the access condition it would be useful to add a last one related to the attempt to set an invalid/unknown access condition (I expect a 422 in such case) in such way we will have a very good coverage for the feature.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/WorkspaceItemRestRepositoryIT.java (hunk 1 above)
Comment (benbosman) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/WorkspaceItemRestRepositoryIT.java]: Yes, we had planned to add tests for verifying the permissions can be set during the bitstream upload, but noticed this was never supported during the file upload.
I've removed the comment now

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/WorkspaceItemRestRepositoryIT.java (hunk 2 above)
Comment (benbosman) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/WorkspaceItemRestRepositoryIT.java]: We had planned to add tests for verifying the permissions can be set during the bitstream upload, but noticed this was never supported during the file upload.
I assumed this was just forgotten during the implementation of the file upload, but if there's no intention to support it, the test will never be relevant
I've removed the test now

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/WorkspaceItemRestRepositoryIT.java (hunk 3 above)
Comment (benbosman) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/WorkspaceItemRestRepositoryIT.java]: The suggested name does indeed make the test much easier to understand.
I've adjusted it now

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/WorkspaceItemRestRepositoryIT.java [hunk 4]
```diff
@@ -3276,65 +3281,6 @@ public void patchUploadTest() throws Exception {
...
-            .andExpect(jsonPath("$",Matchers.allOf(
-                    hasJsonPath("$.sections.upload.files[0].accessConditions[0].name",
-                             is("embargoedWithGroupSelect")),
-                    hasJsonPath("$.sections.upload.files[0].accessConditions[0].groupUUID",
-                             is(embargoedGroup1.getID().toString()))
-                    )));
-    }
-
     @Test
     public void patchUploadRemoveAccessConditionTest() throws Exception {
```
[the remark this answers is not in the document]
Comment (benbosman) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/WorkspaceItemRestRepositoryIT.java]: I've updated the name of the test

**Review by tdonohue — APPROVED** [review]
@benbosman : This is looking good to me now.  However, it cannot be merged (obviously) until we get this rebased to resolve the small merge conflict.  Thanks!

**Review by abollini — APPROVED** [review]
I'm ok with that as well as we have fired a ticket to track future work to support policies for different groups

**Pull request merged**